### PR TITLE
Rover: BalanceBot control improvements

### PR DESCRIPTION
**Pull request by rmackay9**
This PR improves the balance bot control:

- uses the motor library's upper and lower limits to stop I-term build-up in the BB's pitch control.  This is consistent with how we stop I-term build up in other layered controllers.  For example, if the throttle hits 100%, there's no point in the higher-level pitch control building up I-term because it won't help.
- AR_AttitudeControl's pitch controller gets a feedforward based upon the average wheel speed (from the wheel encoders) is added to the output throttle.  This is the biggest change in terms of improving the balancebot's ability to move at higher speeds.
- new WheelRateControl class which uses two PID objects and rate feedback from the wheel encoders to better control the vehicle's wheel speed.  This is optional but when used it should help get rid of non-linear response in the ESC/motors and reduce the effect of the deadzone in the ESCs.
- increase wheel encoder updates from 20hz to 50hz

Tuning a balance bot is made a little easier with these changes:

- DataFlash's PID_Info structure and the AC_PID class allow storing the "actual" replacing the unused "AFF".  This allows the actual rate (or angle or whatever) used at the time of the PID calculation to be recorded.  This makes reporting more accurate and easier.  This should be extended to existing PID objects as well but that can be done as a separate PR.
- add reporting of the Wheel Rate Control PID objects to the GCS in real-time

This code has been used on a real vehicle as [shown in this video](https://www.youtube.com/watch?v=hbPzTOfn3EA&t=21s) but there are a few things we should check before merging:

- does the EKF's position estimation based on wheel encoders work at the new 50hz update rate?  We need to ask @priseborough and/or test on a real vehicle.
- is the WheelRateControl worth the added complexity?  We should disable this control and see if we can still tune up the vehicle to drive as well as it does now.

We may want to eventually add a filter to the wheel-encoder-to-throttle feed-forward but this is not a requirement for merging.

### CODE REVIEW

File: libraries/AP_WheelEncoder/AP_WheelEncoder.cpp [hunk 1]
```diff
@@ -253,6 +255,23 @@ float AP_WheelEncoder::get_distance(uint8_t instance) const
...
+        return 0.0f;
+    }
+
+    // protect against divide by zero
+    if ((state[instance].dt_ms == 0) || _counts_per_revolution[instance] == 0) {
+        return 0;
+    }
+
+    // calculate delta_angle (in radians) per second
+    return M_2PI * state[instance].dist_count_change / _counts_per_revolution[instance] / (state[instance].dt_ms / 1000.0f);
```
Comment (tridge) [inline libraries/AP_WheelEncoder/AP_WheelEncoder.cpp]: parantheses needed

File: libraries/AP_WheelEncoder/AP_WheelRateControl.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+extern const AP_HAL::HAL& hal;
+
+const AP_Param::GroupInfo AP_WheelRateControl::var_info[] = {
+    // @Param: _ENABLE
+    // @DisplayName: Wheel rate control enable/disable
+    // @Description: Enable or disable wheel rate control
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Standard
+    AP_GROUPINFO_FLAGS("_ENABLE", 0, AP_WheelRateControl, _enabled, 0, AP_PARAM_FLAG_ENABLE),
```
Comment (tridge) [inline libraries/AP_WheelEncoder/AP_WheelRateControl.cpp]: don't use index 0 on new param tables

**Comment by rmackay9** [comment]
I'll merge this to master after it passes Travis unless there are any other comments..

**Comment by rmackay9** [comment]
I can't get semaphore to pass but I'm merging anyway because the error doesn't seem related to what I've changed.

**Pull request merged**